test/hotspot/jtreg/compiler/c2/irTests/TestVectorizationMismatchedAccess.java

```diff
@@ -0,0 +1,184 @@
+/*
+ * Copyright (c) 2023, Red Hat, Inc. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package compiler.c2.irTests;
+
+import compiler.lib.ir_framework.*;
+import jdk.test.lib.Utils;
+import jdk.internal.misc.Unsafe;
+import java.util.Objects;
+import java.util.Random;
+
+/*
+ * @test
+ * @modules java.base/jdk.internal.misc
+ * @library /test/lib /
+ * @run driver compiler.c2.irTests.TestVectorizationMismatchedAccess
+ */
+
+public class TestVectorizationMismatchedAccess {
```

Comment:
Could you have a much simpler example, just with two int-arrays, where one allows vectorization and one does not. We still want there to be some vectorization. Maybe like this, maybe you have to play around with it a bit:
```
// make sure arr1 and arr2 are in different slices
for (int i ...) {
  arr1[i] = arr1[i-2] + 5; // do not vectorize, cyclic dependency
  arr2[i] = arr2[i] + 5; // should vectorize
}
```

Reply:
I agree that we should have such a test if we don't have one already but does it belong to that bug? That bug changes the behavior with mismatched accesses so it feels strange to add an unrelated test.` testByteByte1() `is a mismatched access test that vectorizes and that's unaffected by my change. I could add another one that doesn't vectorize because of a dependency with a similar mismatched access. What do you think?

Reply:
Well, my idea was this:
You added a new feature, where we now have a more fine-grained logic. Before, we distinguished by "velt type". Now we distinguish by "memory slice".
Before, this change, my `arr1` and `arr2` would have had the same "velt type", and if one is rejected from vectorization, so is the other. Now they are handled separately, right?

Reply:
I agree with Emanuel here. We need negative tests here too - correctly reject vectorizations which may cause cyclic dependency bugs. For example, passing the same array as 2 arguments (as in `testByteByte1()`) but with different offset to create cyclic dependency.

Reply:
I understand the code better now. I believe that two arrays of the same basic type never land in two separate memory sliced, but always the same. I discussed this with @rwestrel  a while ago. The reason is that EA does not give arrays their own slice, it determines the array to be NSR (not scalar replaceable) if it is ever used in a loop. That is of course a shame, and I hope we can work on this in the future (maybe have some analysis in SuperWord that refines the memory slices, and thus removes unnecessary dependencies). But as @rwestrel  says not the topic of this Bug.
The goal of this thread was to find an example where @rwestrel 's first commit made a worse decision than before. And I found another one after wards (https://github.com/openjdk/jdk/pull/12440#issuecomment-1425919095). @rwestrel now updated the fix, and this specific test should no longer fail.
@vnkozlov adding cyclic dependency tests before fixing https://github.com/openjdk/jdk/pull/12350 will be difficult. Maybe we need to fix both bugs first, and then once they are both integrated add more tests?